lightning/backend/local.go

```diff
@@ -1159,24 +1251,50 @@ func (local *local) ImportEngine(ctx context.Context, engineUUID uuid.UUID) erro
 	return nil
 }
 
-func (local *local) CleanupEngine(ctx context.Context, engineUUID uuid.UUID) error {
+func (local *local) ResetEngine(ctx context.Context, engineUUID uuid.UUID) error {
+	// the only way to reset the engine + reclaim the space is to delete and reopen it 🤷
+	exists, err := local.cleanupEngineIfExists(ctx, engineUUID)
+	if err != nil || !exists {
+		return err
+	}
+	return local.OpenEngine(ctx, engineUUID)
+}
+
+// cleanupEngineIfExists returns if the engine with given UUID exists. If the
+// engine already exists, it will be deleted.
+func (local *local) cleanupEngineIfExists(ctx context.Context, engineUUID uuid.UUID) (exists bool, err error) {
 	// release this engine after import success
 	engineFile, ok := local.engines.Load(engineUUID)
-	if ok {
-		localEngine := engineFile.(*LocalFile)
-		err := localEngine.Close()
-		if err != nil {
-			return err
-		}
-		err = localEngine.Cleanup(local.localStoreDir)
-		if err != nil {
-			return err
-		}
-		local.engines.Delete(engineUUID)
-	} else {
+	if !ok {
 		log.L().Warn("could not find engine in cleanupEngine", zap.Stringer("uuid", engineUUID))
+		return false, nil
 	}
-	return nil
+
+	localEngine := engineFile.(*LocalFile)
+	localEngine.lock()
+	defer localEngine.unlock()
+
+	// since closing the engine causes all subsequent operations on it panic,
+	// we make sure to delete it from the engine map before calling Close().
+	// (note that Close() returning error does _not_ mean the pebble DB
+	// remains open/usable.)
+	local.engines.Delete(engineUUID)
```

Comment:
could this `Delete` happen at same when index engine is flushing in 
https://github.com/pingcap/tidb-lightning/blob/463e7289fcbc412a49e50b307e3b33d30da7f218/lightning/restore/restore.go#L1160
? so flush may access a closed engine. we might need a lock to protect `local.engines`, or extend `LocalFile.importMutex` to protect this close action.